src/Datasource/EntityInterface.php

```diff
@@ -21,30 +21,146 @@
 {
 
+    /**
+     * Sets hidden properties.
+     *
+     * @param array $properties An array of properties to hide from array exports.
+     * @param bool $merge Merge the new properties with the existing. By default false.
+     * @return $this
+     */
+    public function setHidden(array $properties, $merge = false);
+
+    /**
+     * Gets the hidden properties.
+     *
+     * @return array
+     */
+    public function getHidden();
+
+    /**
+     * Sets the virtual properties on this entity.
+     *
+     * @param array $properties An array of properties to treat as virtual.
+     * @param bool $merge Merge the new properties with the existing. By default false.
+     * @return $this
+     */
+    public function setVirtual(array $properties, $merge = false);
+
+    /**
+     * Gets the virtual properties on this entity.
+     *
+     * @return array
+     */
+    public function getVirtual();
+
+    /**
+     * Sets the dirty status of a single property.
+     *
+     * @param string $property the field to set or check status for
+     * @param bool $isDirty true means the property was changed, false means
+     * it was not changed
+     * @return $this
+     */
+    public function setDirty($property, $isDirty);
+
+    /**
+     * Checks if the entity is dirty or if a single property of it is dirty.
+     *
+     * @param string|null $property The field to check the status for. Null for the whole entity.
+     * @return bool Whether the property was changed or not
+     */
+    public function isDirty($property = null);
+
+    /**
+     * Returns all validation errors.
+     *
+     * @return array
+     */
+    public function getErrors();
+
+    /**
+     * Returns validation errors of a field
+     *
+     * @param string $field Field name to get the errors from
+     * @return array
+     */
+    public function getError($field);
+
+    /**
+     * Sets error messages to the entity
+     *
+     * @param array $fields The array of errors to set.
+     * @param bool $overwrite Whether or not to overwrite pre-existing errors for $fields
+     * @return $this
+     */
+    public function setErrors(array $fields, $overwrite = false);
+
+    /**
+     * Sets errors for a single field
+     *
+     * @param string $field The field to get errors for, or the array of errors to set.
+     * @param string|array $errors The errors to be set for $field
+     * @param bool $overwrite Whether or not to overwrite pre-existing errors for $field
+     * @return $this
+     */
+    public function setError($field, $errors, $overwrite = false);
+
+    /**
+     * Stores whether or not a property value can be changed or set in this entity.
+     *
+     * @param string|array $property single or list of properties to change its accessibility
+     * @param bool $set true marks the property as accessible, false will
+     * mark it as protected.
+     * @return $this
+     */
+    public function setAccess($property, $set);
```

Comment:
Should we maybe stop using `string|array $property` mix, and instead offer a separate method for single property, e.g. `setFieldAccess()` or `setAccessForField()`?
This would make typehinting possible. But no big deal, maybe this is more convenient to use in its current form.